File: sdk/java/src/main/java/com/xxxxteam/tempmail/providers/TempMailFyi.java
package com.xxxxteam.tempmail.providers;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.xxxxteam.tempmail.Email;
import com.xxxxteam.tempmail.EmailInfo;
import com.xxxxteam.tempmail.HttpClient;
import com.xxxxteam.tempmail.HttpResult;
import com.xxxxteam.tempmail.Json;
import com.xxxxteam.tempmail.Normalizer;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * temp-mail.fyi 渠道 — https://temp-mail.fyi
 * GET / 获取 PHPSESSID + CSRF，POST /api/generate_email.php 创建，
 * POST /api/get_emails.php 收信。token 使用 base64 编码，前缀 "tmf1:"。
 */
public final class TempMailFyi {

    private static final String BASE_URL = "https://temp-mail.fyi";
    private static final String TOK_PREFIX = "tmf1:";
    private static final String UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
            + "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0";

    private TempMailFyi() {
    }

    private static Map<String, String> apiHeaders(String csrf, String cookie) {
        Map<String, String> h = new LinkedHashMap<>();
        h.put("User-Agent", UA);
        h.put("Accept", "application/json, text/javascript, */*; q=0.01");
        h.put("Accept-Language", "en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7");
        h.put("Content-Type", "application/json");
        h.put("X-CSRF-Token", csrf);
        h.put("X-Requested-With", "XMLHttpRequest");
        h.put("Referer", BASE_URL + "/");
        h.put("Cookie", cookie);
        return h;
    }

    private static String encodeToken(String csrf, String cookie) {
        String raw = "{\"t\":\"" + csrf + "\",\"c\":\"" + cookie.replace("\"", "\\\"") + "\"}";
        return TOK_PREFIX + ProviderUtil.base64Encode(raw.getBytes(StandardCharsets.UTF_8));
    }

    private static String[] decodeToken(String token) {
        if (token == null || !token.startsWith(TOK_PREFIX)) {
            throw new RuntimeException("tempmail-fyi: 无效的 token");
        }
        byte[] decoded = ProviderUtil.base64Decode(token.substring(TOK_PREFIX.length()));
        JsonElement obj = Json.parse(new String(decoded, StandardCharsets.UTF_8));
        String csrf = Json.str(obj, "t").trim();
        String cookie = Json.str(obj, "c").trim();
        if (csrf.isEmpty()) {
            throw new RuntimeException("tempmail-fyi: token 中缺少 CSRF");
        }
        return new String[]{csrf, cookie};
    }

    /**
     * 创建临时邮箱。
     *
     * @return 邮箱信息
     */
    public static EmailInfo generate() {
        Map<String, String> pageH = new LinkedHashMap<>();
        pageH.put("User-Agent", UA);
        pageH.put("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8");
        HttpResult resp = HttpClient.get(BASE_URL + "/", pageH);
        resp.ensureSuccess();
        String cookie = ProviderUtil.extractAllCookies(resp);
        // 提取 CSRF: csrfToken" value="xxx"
        String csrf = ProviderUtil.regexExtract(resp.getBody(), "csrfToken\"\\s*value=\"([^\"]+)\"");
        if (csrf.isEmpty()) {
            throw new RuntimeException("tempmail-fyi: 未能从首页提取 CSRF token");
        }
        // POST /api/generate_email.php
        HttpResult resp2 = HttpClient.post(BASE_URL + "/api/generate_email.php", "{}",
                "application/json", apiHeaders(csrf, cookie));
        resp2.ensureSuccess();
        JsonElement data = Json.parse(resp2.getBody());
        JsonElement successEl = data.getAsJsonObject().get("success");
        if (successEl == null || !successEl.getAsBoolean()) {
            throw new RuntimeException("tempmail-fyi: 创建邮箱失败");
        }
        String address = Json.str(data, "email_address").trim();
        if (address.isEmpty() || !address.contains("@")) {
            throw new RuntimeException("tempmail-fyi: 邮箱地址无效");
        }
        return new EmailInfo("tempmail-fyi", address, encodeToken(csrf, cookie), null, null);
    }

    /**
     * 获取邮件列表。
     *
     * @param token base64 编码的认证信息
     * @param email 邮箱地址
     * @return 邮件列表
     */
    public static List<Email> getEmails(String token, String email) {
        String address = (email != null ? email : "").trim();
        if (address.isEmpty()) {
            throw new RuntimeException("tempmail-fyi: 邮箱地址为空");
        }
        String[] parts = decodeToken(token);
        String csrf = parts[0];
        String cookie = parts[1];
        String body = "{\"email_address\":\"" + address + "\"}";
        HttpResult resp = HttpClient.post(BASE_URL + "/api/get_emails.php", body,
                "application/json", apiHeaders(csrf, cookie));
        resp.ensureSuccess();
        JsonElement data = Json.parse(resp.getBody());
        JsonElement successEl = data.getAsJsonObject().get("success");
        if (successEl == null || !successEl.getAsBoolean()) {
            return new ArrayList<>();
        }
        JsonArray emails = Json.arr(data, "emails");
        if (emails == null || emails.size() == 0) {
            return new ArrayList<>();
        }
        List<Email> result = new ArrayList<>();
        for (JsonElement item : emails) {
            if (item == null || !item.isJsonObject()) continue;
            result.add(Normalizer.normalizeEmail(Json.toDict(item), address));
        }
        return result;
    }
}

File: sdk/kotlin/src/main/kotlin/com/xxxxteam/tempmail/providers/TempMailFyi.kt
package com.xxxxteam.tempmail.providers

import com.xxxxteam.tempmail.Email
import com.xxxxteam.tempmail.EmailInfo
import com.xxxxteam.tempmail.Http
import com.xxxxteam.tempmail.Normalize
import com.xxxxteam.tempmail.Provider
import kotlinx.serialization.json.*
import java.nio.charset.StandardCharsets
import java.util.Base64

/**
 * temp-mail.fyi 渠道（temp-mail.fyi）。
 *
 * GET / 取 PHPSESSID + CSRF，POST /api/generate_email.php 创建，
 * POST /api/get_emails.php 收信。token 使用 base64 编码，前缀 "tmf1:"。
 */
object TempMailFyi : Provider {

    private const val BASE_URL = "https://temp-mail.fyi"
    private const val TOK_PREFIX = "tmf1:"
    private const val UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 " +
        "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0"

    private fun apiHeaders(csrf: String, cookie: String) = mapOf(
        "User-Agent" to UA,
        "Accept" to "application/json, text/javascript, */*; q=0.01",
        "Accept-Language" to "en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7",
        "Content-Type" to "application/json",
        "X-CSRF-Token" to csrf,
        "X-Requested-With" to "XMLHttpRequest",
        "Referer" to "$BASE_URL/",
        "Cookie" to cookie,
    )

    /** 创建临时邮箱。 */
    override suspend fun generate(): EmailInfo {
        val resp = ProviderUtil.httpGet("$BASE_URL/", mapOf(
            "User-Agent" to UA,
            "Accept" to "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"))
        resp.ensureSuccess()
        val cookie = extractAllCookies(resp.setCookies)
        val csrf = Regex("csrfToken\"\\s*value=\"([^\"]+)\"").find(resp.body)?.groupValues?.get(1) ?: ""
        if (csrf.isBlank()) throw RuntimeException("tempmail-fyi: 未能从首页提取 CSRF token")
        val resp2 = ProviderUtil.httpPost("$BASE_URL/api/generate_email.php", "{}",
            "application/json", apiHeaders(csrf, cookie))
        resp2.ensureSuccess()
        val data = Http.json.parseToJsonElement(resp2.body) as? JsonObject
            ?: throw RuntimeException("tempmail-fyi: 创建邮箱失败")
        if ((data["success"] as? JsonPrimitive)?.booleanOrNull != true) {
            throw RuntimeException("tempmail-fyi: 创建邮箱失败")
        }
        val address = jsonStr(data, "email_address").trim()
        if (address.isBlank() || !address.contains("@")) throw RuntimeException("tempmail-fyi: 邮箱地址无效")
        return EmailInfo(email = address, channel = "tempmail-fyi", token = encodeToken(csrf, cookie))
    }

    /** 获取邮件列表。 */
    override suspend fun getEmails(info: EmailInfo): List<Email> {
        val address = info.email.trim()
        if (address.isBlank()) throw RuntimeException("tempmail-fyi: 邮箱地址为空")
        val (csrf, cookie) = decodeToken(info.token)
        val body = buildJsonObject { put("email_address", address) }.toString()
        val resp = ProviderUtil.httpPost("$BASE_URL/api/get_emails.php", body,
            "application/json", apiHeaders(csrf, cookie))
        resp.ensureSuccess()
        val data = Http.json.parseToJsonElement(resp.body) as? JsonObject ?: return emptyList()
        if ((data["success"] as? JsonPrimitive)?.booleanOrNull != true) return emptyList()
        val emails = data["emails"] as? JsonArray ?: return emptyList()
        return emails.mapNotNull { it as? JsonObject }.map { Normalize.fromJson(it, address) }
    }

    /** 编码 token（base64，前缀 tmf1:）。 */
    private fun encodeToken(csrf: String, cookie: String): String {
        val raw = buildJsonObject { put("t", csrf); put("c", cookie) }.toString()
        return TOK_PREFIX + Base64.getEncoder().encodeToString(raw.toByteArray(StandardCharsets.UTF_8))
    }

    /** 解码 token，返回 [csrf, cookie]。 */
    private fun decodeToken(token: String): Pair<String, String> {
        if (!token.startsWith(TOK_PREFIX)) throw RuntimeException("tempmail-fyi: 无效的 token")
        val decoded = Base64.getDecoder().decode(token.substring(TOK_PREFIX.length))
        val obj = Http.json.parseToJsonElement(String(decoded, StandardCharsets.UTF_8)) as? JsonObject
            ?: throw RuntimeException("tempmail-fyi: 无效的 token")
        val csrf = jsonStr(obj, "t").trim()
        val cookie = jsonStr(obj, "c").trim()
        if (csrf.isEmpty()) throw RuntimeException("tempmail-fyi: token 中缺少 CSRF")
        return csrf to cookie
    }

    /** 提取所有 Set-Cookie 的 name=value 串。 */
    private fun extractAllCookies(setCookies: List<String>): String =
        setCookies.map { it.split(";")[0].trim() }.filter { it.contains("=") }.joinToString("; ")
}

/** 从 JSON 对象读取字符串字段，缺失返回空串（文件私有）。 */
private fun jsonStr(obj: JsonObject, key: String): String =
    (obj[key] as? JsonPrimitive)?.contentOrNull ?: ""
